File: 101.symmetric-tree.js
/*
 * @lc app=leetcode id=101 lang=javascript
 *
 * [101] Symmetric Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function (root) {
	if (root == null) return true;
	return dfs(root.left, root.right);
};

function dfs(left, right) {
	if (left == null && right == null) return true;
	if (left == null || right == null) return false;
	if (left.val != right.val) return false;
	return dfs(left.left, right.right) && dfs(left.right, right.left)
}
// @lc code=end

File: 101.symmetric-tree.py
#
# @lc app=leetcode id=101 lang=python3
#
# [101] Symmetric Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if root is None:
            return True
        return self.isSymmetric_check(root, root)

    def isSymmetric_check(self, left, right) -> bool:
        if left and right:
            return (
                left.val == right.val
                and self.isSymmetric_check(left.left, right.right)
                and self.isSymmetric_check(left.right, right.left)
            )
        else:
            return left == right


# @lc code=end